src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CodeFixes/AttributeUpgradeCodeFixer.cs

```diff
@@ -0,0 +1,153 @@
+
+            // Remove the attribute or, if it's the only attribute in the attribute list, remove the attribute list
+            if ((node.Parent is CSSyntax.AttributeListSyntax csList && csList.Attributes.Count == 1)
+                || (node.Parent is VBSyntax.AttributeListSyntax vbList && vbList.Attributes.Count == 1))
+            {
+                // We want to remove trivia with the attribute list, but we want to *keep* leading end of line trivia
+                // so that we don't remove blank lines before the property that had the attribute list on it.
+                // To do that, we first remove any leading trivia except new lines and then remove the node keeping
+                // leading trivia.
+                var trimmedList = node.Parent!.WithLeadingTrivia(node.Parent!.GetLeadingTrivia().Where(t => t.IsKind(CS.SyntaxKind.EndOfLineTrivia) || t.IsKind(VB.SyntaxKind.EndOfLineTrivia)));
+                editor.ReplaceNode(node.Parent, trimmedList);
+                editor.RemoveNode(trimmedList, SyntaxRemoveOptions.KeepLeadingTrivia);
+            }
+            else
+            {
+                editor.RemoveNode(node, SyntaxRemoveOptions.KeepNoTrivia);
+            }
+
+            return editor.GetChangedDocument();
+        }
+
+        private static async Task<Document> ReplaceAttributeAsync(Document document, SyntaxNode node, string attributeType, CancellationToken cancellationToken)
+        {
+            var generator = SyntaxGenerator.GetGenerator(document);
+
+            // Create an updated Attribute node
+            var updatedNode = node switch
+            {
+                CSSyntax.AttributeSyntax csAttribute => (SyntaxNode)csAttribute.WithName((CSSyntax.NameSyntax)QualifiedNameBuilder.BuildQualifiedNameSyntax(generator, attributeType)),
+                VBSyntax.AttributeSyntax vbAttribute => vbAttribute.WithName((VBSyntax.NameSyntax)QualifiedNameBuilder.BuildQualifiedNameSyntax(generator, attributeType)),
+                _ => throw new InvalidOperationException($"Unexpected syntax node type: {node.GetType().FullName}")
+            };
+            updatedNode = updatedNode.WithAdditionalAnnotations(Simplifier.Annotation);
+
+            // Split the new idenfitier into namespace and name components
+            var namespaceDelimiterIndex = attributeType.LastIndexOf('.');
+            var qualifier = namespaceDelimiterIndex >= 0
+                ? attributeType.Substring(0, namespaceDelimiterIndex)
+                : null;
+
+            // The simplifier does not recognize using statements within namespaces, so this
+            // checks whether the necessary using statement is present or not and adds the
+            // AddImportsAnnotation only if it's absent.
+            if (qualifier is not null && !node.HasAccessToNamespace(qualifier))
+            {
+                updatedNode = updatedNode.WithAdditionalAnnotations(Simplifier.AddImportsAnnotation);
```

Comment:
This is a new API I believe. We had been targeting support in VS 16.8 (which supports Roslyn 3.8) due to the requirement for .NET 5. However, looks like we've updated to 3.9. Was this intentional? We should update a requirements then.

Reply:
I just checked and this is in 3.8.0. Good point about not wanting to be on a newer version, though. I opened a separate PR to roll that back (#653).